datalad/plugin/addurls.py

```diff
@@ -0,0 +1,654 @@
+    missing_value : str, optional
+        When an empty string is encountered, use this value instead.
+
+    Examples
+    --------
+    Consider a file "avatars.csv" that contains
+
+        who,ext,link
+        neurodebian,png,https://avatars3.githubusercontent.com/u/260793
+        datalad,png,https://avatars1.githubusercontent.com/u/8927200
+
+    To download each link into a file name composed of the 'who' and
+    'ext' fields, we could run
+
+        $ datalad plugin -d avatar_ds addurls url_file=avatars.csv
+          url_format='{link}' filename_format='{who}.{ext}' fast=True
+
+    The '-d avatar_ds' is used to create a new dataset in
+    "$PWD/avatar_ds".
+
+    If we were already in a dataset and wanted to create a new
+    subdataset in an "avatars" subdirectory, we could use "//" in the
+    `filename_format` argument:
+
+        $ datalad plugin addurls url_file=avatars.csv
+          url_format='{link}' filename_format='avatars//{who}.{ext}'
+          fast=True
+
+    Note
+    ----
+    For users familiar with 'git annex addurl': A large part of this
+    plugin's functionality can be viewed as transforming data from
+    `url_file` into a "url filename" format that fed to 'git annex
+    addurl --batch --with-files'.
+    """
+    import logging
+    import os
+
+    from datalad.distribution.dataset import Dataset
+    from datalad.interface.results import get_status_dict
+    import datalad.plugin.addurls as me
+    from datalad.support.annexrepo import AnnexRepo
+    from datalad.support.exceptions import AnnexBatchCommandError
+    from datalad.utils import assure_list
+    from datalad.ui import ui
+
+    lgr = logging.getLogger("datalad.plugin.addurls")
+
+    meta = assure_list(meta)
+
+    if url_file is None:
+        # `url_file` is not a required argument in `dlplugin` because
+        # the argument before it, `dataset`, needs to be optional to
+        # support the creation of new datasets.
+        yield get_status_dict(action="addurls",
+                              ds=dataset,
+                              status="error",
+                              message="Must specify url_file argument")
+        return
+
+    if dataset.repo and not isinstance(dataset.repo, AnnexRepo):
+        yield get_status_dict(action="addurls",
+                              ds=dataset,
+                              status="error",
+                              message="not an annex repo")
+        return
+
+    if input_type == "ext":
+        extension = os.path.splitext(url_file)[1]
+        input_type = "json" if extension == ".json" else "csv"
+
+    with open(url_file) as fd:
+        rows, subpaths = me.extract(fd, input_type,
+                                    filename_format, url_format,
+                                    exclude_autometa, meta,
+                                    missing_value)
+
+    all_files = [row["filename"] for row in rows]
+    if len(all_files) != len(set(all_files)):
```

Comment:
have you tried on that ultimate target usecase document 
https://github.com/OpenNeuroLab/metasearch/issues/14#issuecomment-367121988
?
I was wondering what would happen if some rows do not have a URL -- wouldn't it also trigger this condition?  feels like it should not be just a comparison by size but throw an exception from the guts if collision happens

Reply:
also I am afraid we cannot escape supporting disposition filename since there is simply no way around -- without it I would not even know what extension to assign e.g. for the following record:
```
ixi,1,582,29.77138946,control,,university or polytechnic degree,white,,,married,go out to full time employment,,Male,ixi,homo-sapiens,,https://files.osf.io/v1/resources/5h7sv/providers/osfstorage/5839b790594d9000421f6ca7,1,yes
```
you could use 
```shell
$> python -c 'from datalad.support.network import get_url_filename as gurl; print(gurl("https://files.osf.io/v1/resources/5h7sv/providers/osfstorage/5839b790594d9000421f6ca7"))'           
IXI582-Guys-1127-T1.nii.gz
```
which would also just use url filename if no disposition entry in the header BUT that elderly one is not working correctly -- it downloads the entire beast ;)  Please fix it up to use our downloaders which do it without downloading the entire beast.  E.g. if we are just to limit ourselves to http without authentication for now:
```
$> python -c 'from datalad.downloaders.http import HTTPDownloader; print HTTPDownloader().get_status("https://files.osf.io/v1/resources/5h7sv/providers/osfstorage/5839b790594d9000421f6ca7")'                                                    
FileStatus(filename='IXI582-Guys-1127-T1.nii.gz', mtime=1480177552, size=8225627)
```
so there you could have your filename quickly (although with the server request).  I guess should happen optionally if `url_filename` is used in the `filename_format`.

let's concentrate on this use case please  so we could crawl it and get our lovely dataset with sensible filenames

Reply:
> also I am afraid we cannot escape supporting disposition filename 

OK, thanks for helping me see the point.

Reply:
> I was wondering what would happen if some rows do not have a URL -- wouldn't it also trigger this condition?

That collision test is for the destination files rather than the URLs.